File: app/src/main/java/com/andb/apps/todo/utilities/ProjectsUtils.java
package com.andb.apps.todo.utilities;

import android.os.AsyncTask;

import com.andb.apps.todo.MainActivity;
import com.andb.apps.todo.databases.ProjectsDatabase;
import com.andb.apps.todo.lists.ProjectList;
import com.andb.apps.todo.objects.Project;
import com.andb.apps.todo.objects.Tasks;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.Random;

public class ProjectsUtils{

    public static void update(Project project, ProjectsDatabase projectsDatabase){
        AsyncTask.execute(() -> projectsDatabase.projectsDao().updateProject(project));

    }

    public static void update(Project project){
        update(project, MainActivity.projectsDatabase);
    }

    public static void update(ProjectsDatabase projectsDatabase){
        update(Current.project(), projectsDatabase);
    }

    public static void update(){
        update(Current.project(), MainActivity.projectsDatabase);
    }

    public static Project projectFromKey(int key){
        for (Project p : Current.allProjects()){
            if (p.getKey()==key){
                return p;
            }
        }
        return null;
    }

    private static Random random = new Random();
    public static int keyGenerator(){
        int key = random.nextInt();
        ArrayList<Integer> keys = new ArrayList<>();
        keys.addAll(Current.keyList());
        for (Project project : Current.allProjects()){
            keys.add(project.getKey());
        }

        while (keys.contains(key)){
            key = random.nextInt();
        }
        return key;
    }
}

File: app/src/main/java/com/andb/apps/todo/utilities/ProjectsUtils.kt
package com.andb.apps.todo.utilities

import android.os.AsyncTask
import com.andb.apps.todo.data.local.Database
import com.andb.apps.todo.databases.projectsDao
import com.andb.apps.todo.databases.tagsDao
import com.andb.apps.todo.databases.tasksDao
import com.andb.apps.todo.data.model.Project
import com.andb.apps.todo.data.model.Tag
import com.andb.apps.todo.data.model.Task
import java.util.*

object ProjectsUtils {

    private val random = Random()

    @JvmOverloads
    @JvmStatic
    fun update(project: Project = Current.project(), database: Database = Current.database()) {
        AsyncTask.execute {
            database.projectsDao().updateProject(project)
        }
    }

    @JvmOverloads
    @JvmStatic
    fun update(tag: Tag, database: Database = Current.database(), async: Boolean = true) {
        if(async) {
            AsyncTask.execute {
                database.tagsDao().updateTag(tag)
            }
        }else{
            database.tagsDao().updateTag(tag)
        }
    }

    @JvmOverloads
    @JvmStatic
    fun update(task: Task, database: Database = Current.database(), async: Boolean = true) {
        if(async) {
            AsyncTask.execute {
                database.tasksDao().updateTask(task)
            }
        }else{
            database.tasksDao().updateTask(task)
        }
    }

    @JvmStatic
    fun remove(tag: Tag){
        AsyncTask.execute {
            for(task in Current.taskListAll()){
                if(task.listTags.contains(tag.id)){
                    task.listTags.remove(tag.id)
                    update(task, async = false)
                }
            }
            for(childHolder in Current.tagListAll()){
                if(childHolder.children.contains(tag.id)){
                    childHolder.children.remove(tag.id)
                    update(childHolder, async = false)
                }
            }
            for(t in Current.tagListAll()){
                if (t.index>tag.index){
                    t.index--
                }
            }

            tagsDao().deleteTag(tag)
        }
    }

    @JvmStatic
    fun remove(task: Task){
        AsyncTask.execute {
            tasksDao().deleteTask(task)
        }
    }


    @JvmStatic
    fun projectFromKey(key: Int): Project? {
        for (p in Current.allProjects()) {
            if (p.id == key) {
                return p
            }
        }
        return null
    }

    @JvmStatic
    fun addProject(name: String, color: Int): Project {
        val project = Project(keyGenerator(), name, color, Current.allProjects().size)
        AsyncTask.execute {
            projectsDao().insertOnlySingleProject(project)
        }

        return project
    }

    @JvmStatic
    fun keyGenerator(): Int {
        var key = random.nextInt()

        while (Current.keyList().contains(key) || key in listOf(0, -1)) {
            key = random.nextInt()
        }
        return key
    }


}
